File: 2019/day06_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Tree = tuple[str, list[Tree]]


def solution(content: list[str]) -> Any:
    edges = [parse(line) for line in content]
    lefts = [left for left, _ in edges]
    rights = [right for _, right in edges]
    root = [item for item in lefts if item not in rights][0]
    return count(tree(root, edges))


def count(tr: Tree):
    def helper(value: int, t: Tree) -> int:
        _, rest = t
        return value + sum(helper(value + 1, r) for r in rest)

    return helper(0, tr)


def tree(root: str, edges: list[tuple[str, str]]) -> Tree:
    return (
        root,
        [tree(b, edges) for a, b in edges if a == root],
    )


def parse(line: str) -> tuple[str, str]:
    parts = line.split(")")
    return (parts[0].strip(), parts[1].strip())


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2019/day06-1.ts
type Edge = {
  from: string;
  to: string;
};

type Tree = {
  label: string;
  subtrees: Tree[];
};

function solution(content: string) {
  const edges = content.split("\n").map(parse);
  const lefts = edges.map(({ from }) => from);
  const rights = edges.map(({ to }) => to);
  const root = lefts.find((left) => !rights.includes(left));
  if (root) return count(tree(root, edges));
}

function count(tr: Tree) {
  function helper(value: number, t: Tree): number {
    const { subtrees } = t;
    return (
      value + subtrees.reduce((acc, cur) => acc + helper(value + 1, cur), 0)
    );
  }
  return helper(0, tr);
}

function tree(label: string, edges: Edge[]): Tree {
  return {
    label,
    subtrees: edges
      .filter(({ from }) => from === label)
      .map(({ to }) => tree(to, edges)),
  };
}

function parse(line: string): Edge {
  const parts = line.split(")");
  return {
    from: parts[0],
    to: parts[1],
  };
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
